src/weighted_list.h

```diff
@@ -17,37 +17,43 @@ template <typename W, typename T> struct weighted_list {
     weighted_list() : total_weight(0) { };
 
     /**
-     * This will add a new object to the weighted list.
+     * This will add a new object to the weighted list. Returns a pointer to
+       the added or updated object, or NULL if weight was zero.
```

Comment:
The documentation is a bit misleading. If weight is 0 (or less), it will return a pointer to the last element, or invoke undefined behavior if the list is currently empty. It will never (reliably) return NULL.